File: app/tools/select_tool.py
from PySide6.QtCore import QPointF, Qt
from PySide6.QtGui import QMouseEvent
from app.tools.base_tool import BaseTool

class SelectMoveTool(BaseTool):
    """
    Select & Move tool for dragging canvas or adjusting position of layers/foreground.
    """

    def __init__(self, canvas):
        super().__init__(canvas)
        self.is_dragging = False
        self.last_pos = None

    def mouse_press(self, img_pos: QPointF, event: QMouseEvent):
        if event.button() == Qt.MouseButton.LeftButton:
            self.is_dragging = True
            self.last_pos = event.position()

    def mouse_move(self, img_pos: QPointF, event: QMouseEvent):
        if self.is_dragging and self.last_pos is not None:
            delta = event.position() - self.last_pos
            self.last_pos = event.position()
            # Pan canvas viewport
            hs = self.canvas.horizontalScrollBar()
            vs = self.canvas.verticalScrollBar()
            hs.setValue(int(hs.value() - delta.x()))
            vs.setValue(int(vs.value() - delta.y()))

    def mouse_release(self, img_pos: QPointF, event: QMouseEvent):
        if event.button() == Qt.MouseButton.LeftButton:
            self.is_dragging = False
            self.last_pos = None

File: Cpp/src/tools/SelectTool.cpp
#include "tools/SelectTool.h"
#include "ui/CanvasView.h"

namespace ImageCut {
namespace Tools {

SelectMoveTool::SelectMoveTool(UI::CanvasView* canvas) : BaseTool(canvas) {}

void SelectMoveTool::mousePress(const QPointF& imgPos, QMouseEvent* event) {
    (void)imgPos;
    if (event->button() == Qt::LeftButton) {
        m_isDragging = true;
        m_lastPos = event->position();
    }
}

void SelectMoveTool::mouseMove(const QPointF& imgPos, QMouseEvent* event) {
    (void)imgPos;
    if (m_isDragging && m_canvas) {
        QPointF delta = event->position() - m_lastPos;
        m_lastPos = event->position();
        m_canvas->panViewport(delta.x(), delta.y());
    }
}

void SelectMoveTool::mouseRelease(const QPointF& imgPos, QMouseEvent* event) {
    (void)imgPos;
    if (event->button() == Qt::LeftButton) {
        m_isDragging = false;
    }
}

} // namespace Tools
} // namespace ImageCut
